src/erldns_zone_cache.erl

```diff
@@ -249,7 +346,73 @@ delete_zone_records(Name) ->
   erldns_storage:select_delete(zone_records, [{{{erldns:normalize_name(Name), '_'}, '_'},[],[true]}]),
   erldns_storage:select_delete(zone_records_typed, [{{{erldns:normalize_name(Name), '_', '_'}, '_'},[],[true]}]).
 
-
+%% @doc Remove zone RRSet
+-spec delete_zone_rrset(binary(), binary(), binary(), integer()) -> any().
+delete_zone_rrset(Name, Digest, RRFqdn, Type) ->
+  delete_zone_rrset(Name, Digest, RRFqdn, Type, 0).
+
+%% @doc Remove zone RRSet
+-spec delete_zone_rrset(binary(), binary(), binary(), integer(), integer()) -> any().
+delete_zone_rrset(ZoneName, Digest, RRFqdn, Type, Counter) ->
+  case find_zone_in_cache(erldns:normalize_name(ZoneName)) of
+    {ok, Zone} -> 
+	  Zone,
+	  CurrentCounter = get_rrset_sync_counter(ZoneName, RRFqdn, Type),
+	  if 
+	    Counter =:= 0 orelse CurrentCounter < Counter -> 
+		  lager:debug("Removing RRSet (~p) with type ~p", [RRFqdn, Type]),
+		  erldns_storage:select_delete(zone_records, [{{{erldns:normalize_name(ZoneName), erldns:normalize_name(RRFqdn)}, '_'},[],[true]}]),
+		  erldns_storage:select_delete(zone_records_typed, [{{{erldns:normalize_name(ZoneName), erldns:normalize_name(RRFqdn), Type}, '_'},[],[true]}]),
+		  % delete RRSet related RRSig records
+		  % get RRSIG records without type covered first
+		  RRSigRecs = filter_rrsig_records_with_type_covered(RRFqdn, Type),
+		  % delete all RRSIGs for FQDN 
+		  erldns_storage:select_delete(zone_records_typed, [{{{erldns:normalize_name(ZoneName), erldns:normalize_name(RRFqdn), ?DNS_TYPE_RRSIG_NUMBER}, '_'},[],[true]}]),
+		  % write back the filtered RRSIG set
+		  lists:map(fun(R) ->
+			erldns_storage:insert(zone_records_typed, R) end,
+			RRSigRecs),
+		  % only write counter if called explicitly with Counter value i.e. different than 0.
+		  % this will not write the counter if called by put_zone_rrset/3 as it will prevent subsequent delete ops
+		  if 
```

Comment:
As above, I generally use case matching whenever possible rather than `if`.

Reply:
Agreed, I will refactor the Ifs.

Reply:
commit: https://github.com/dnsimple/erldns/pull/100/commits/1788d173d41c09fc5ab6780a5d029d62c5d2cd94 refactors the ifs to case of expressions